pypeit/pypeit.py

```diff
@@ -4,157 +4,505 @@
+            if not os.path.exists(newdir):
+                os.mkdir(newdir)
+            # Now the file
+            pypeit_file = os.path.join(newdir, root+setup+'.pypeit')
+            # Modify parlines
+            for kk in range(len(cfg_lines)):
+                if 'sortroot' in cfg_lines[kk]:
+                    cfg_lines[kk] = '    sortroot = {0}'.format(root+setup)
+
+            make_pypeit_file(pypeit_file, self.spectrograph.spectrograph, [], cfg_lines=cfg_lines,
+                             setup_lines=setup_lines, sorted_files=sorted_files, paths=paths)
+            print("Wrote {:s}".format(pypeit_file))
+
+    def build_qa(self):
+        qa.gen_mf_html(self.pypeit_file)
         qa.gen_exp_html()
-    else:
-        msgs.error('Data reduction failed with status ID {0:d}'.format(status))
-
-    # Capture the end time and print it to user
-    tend = time.time()
-    codetime = tend-tstart
-    if codetime < 60.0:
-        msgs.info('Data reduction execution time: {0:.2f}s'.format(codetime))
-    elif codetime/60.0 < 60.0:
-        mns = int(codetime/60.0)
-        scs = codetime - 60.0*mns
-        msgs.info('Data reduction execution time: {0:d}m {1:.2f}s'.format(mns, scs))
-    else:
-        hrs = int(codetime/3600.0)
-        mns = int(60.0*(codetime/3600.0 - hrs))
-        scs = codetime - 60.0*mns - 3600.0*hrs
-        msgs.info('Data reduction execution time: {0:d}h {1:d}m {2:.2f}s'.format(hrs, mns, scs))
-
-    return status
+
+    def calibrate_one(self, sci_ID, det):
+        pass
+
+    def extract_all(self, reuse=False):
+
+        self.tstart = time.time()
+        std_dict = {}
+        all_sci_ID = self.fitstbl['sci_ID'].data[self.fitstbl['science']]  # Binary system: 1,2,4,8, etc.
+        numsci = len(all_sci_ID)
+        basenames = [None]*numsci  # For fluxing at the very end
+
+        # Check par
+        required = [ 'rdx', 'calibrations', 'scienceframe', 'scienceimage', 'flexure', 'fluxcalib' ]
+        can_be_None = [ 'flexure', 'fluxcalib' ]
+        self.par.validate_keys(required=required, can_be_None=can_be_None)
+
+        for sci_ID in all_sci_ID:
+            sci_dict = self.extract_exposure(sci_ID, reuse=reuse)
+            self.save_exposure(sci_ID, sci_dict)
+
+        self.print_end_time()
+
+    def extract_exposure(self, sci_ID, reuse=False):
+        self.sci_ID = sci_ID
+
+
+        if reuse:
+            self.par['calibrations']['masters'] = 'reuse'
+
+        sci_dict = OrderedDict()  # This needs to be ordered
+        sci_dict['meta'] = {}
+        sci_dict['meta']['vel_corr'] = 0.
+        #
+        scidx = np.where((self.fitstbl['sci_ID'] == sci_ID) & self.fitstbl['science'])[0][0]
+        msgs.info("Reducing file {0:s}, target {1:s}".format(self.fitstbl['filename'][scidx],
+                                                             self.fitstbl['target'][scidx]))
+
+        # Loop on Detectors
+        for kk in range(self.spectrograph.ndet):
```

Comment:
How will you deal with a situation like LRIS red where the number of detectors is changing? I guess the spectrograph class needs to know something about the file then. The fundamental problem is we cannot know ndet until we've seen a file.